### ISSUE TITLE
Add MetalLB integration, documentation, and support to Rancher

### ISSUE THREAD

**Issue by Tejeev** [rancher/rancher#27627]
**What kind of request is this (question/bug/enhancement/feature request):**
RFE

**Request**
It would be very helpful to add integration of MetalLB to Rancher, flesh out the documentation, and add it to Rancher Support matrix.  The ability to deploy MetalLB via RKE would be very helpful.

**Use Case**
We are deploying in a factory environment (industry 4.0) with an on-prem Rancher installation running in VMware vSphere. 

We are able to utilize nginx-ingress to route all of our L7 traffic successfully and scalably. We have a firewall cluster at the heart of our network which we use for LB functionality, once the traffic hits Rancher we use L7 logic to route to workloads. This gives us a nice one-to-many north-south solution.

A lot of traffic and workloads that we use in industry 4.0 (majority) are not web-based. A lot of it is (very) painful legacy and focused on TCP connections (sometimes not even understanding DNS), for example, OPC-UA clients running in Rancher. We can use nodeport to expose this traffic for north-south except that this introduces a many-to-many connection problem for L4 north-south which is not scalable and not agile.

It seems like MetalLB is the best solution for our L4 traffic as it allows us to bypass the network load balancer and create a software-defined one-to-many solution for us. In researching this I'm finding a lack of Rancher documentation on MetalLB (to use alongside nginx-ingress).

gz#10718
gz#11274
gz#12486

**Comment by fabiorauber** [comment]
I'll post another use case:

I was studying the documentation for installation of Rancher in HA mode, inside an RKE cluster. The docs states that an External Load Balancer is needed, either a L4 or L7 one. Although there are instructions on how to provision a L7 LB using Nginx (https://rancher.com/docs/rancher/v2.x/en/installation/options/nginx/), the process is not automated. In addition, there are no instructions provided to install a L4 load balancer, which is the recommended Load Balancer type, as taught in Rancher Academy videos.

A great option would be to provision MetalLB via RKE. Another good option would be to document the installation of MetalLB to serve as a Load Balancer for the Rancher HA Cluster.

**Comment by Tejeev** [comment]
Agreed that provisioning metallb with ingress via RKE would be the end goal for our factory environment as well.

**Comment by leflambeur** [comment]
Adding to @Tejeev 

> So I'm using the metallb.universe.tf/allow-shared-ip to just expose 1 virtual IP and use port mappings for different workloads. As it currently stands I can create a fault situation by using duplicate port mappings. Because MetalLB is not integrated into Rancher but is a loose component Rancher is not aware that metalLB is logging this as an error from the controller component ({"caller":"service.go:106","error":"no available IPs","msg":"IP allocation failed","op":"allocateIP","service":"spielerij/helloworld2-loadbalancer","ts":"2020-06-19T08:00:28.792568466Z"} ) and so the LB continues to be stuck on 'pending' status in Rancher.
> 
> Ideally Rancher would display the error, that could be one of the value adds of the integration

**Comment by stefanlasiewski** [comment]
We also have made the same request to Rancher. 

I do have concerns about the viability of MetalLB, since it is still Beta software and only has one main developer (Who has been doing a great job, [but who is limited in terms of how much effort they can put in](https://metallb.universe.tf/concepts/maturity/#developers)).

**Comment by asemen2608** [comment]
**What kind of request is this (question/bug/enhancement/feature request):**
RFE

**Request**
Out the box Metallb support in rancher

**Use Case**
For on-prem becoming defacto for loadbalance support is metallb, can we request out the box support for metallb, with the ability to have externaldns in coredns and certmanager. Out the box support ability to configure it at the cluster configuration level, inbuilt metrics/monitoring/alerts, and panels in the rancher UI (in a similar way to nginx is out the box)

gz#12486

**Comment by asemen2608** [comment]
Please provide a date/eta when RFE will be implemented.
Customer is asking.

**Comment by lusien88** [comment]
I agree with @Tejeev ! It would be grate to have MetalLB deployed directly with the nginx-ingress-controller load-balancer type.

**Comment by stefanlasiewski** [comment]
Support for MetalLB as a L4 Loadbalancer would provide better integration with [kubernetes-sigs/external-dns](https://github.com/kubernetes-sigs/external-dns) services, and would greatly simplify the process of exposing non-HTTP services that are not well supported by the Nginx Ingress.

**Comment by immanuelfodor** [comment]
@fabiorauber If you want to get rid of the external load balancer, I recommend checking out Kube Karp that provides a floating virtual IP to the nodes based on the CARP protocol, so you can be sure that your floating IP is always pointing to a node that is up and running: https://github.com/immanuelfodor/kube-karp

Say you have nodes with IPs 192.168.1.101, 192.168.1.102, 192.168.1.103, then you can have an IP like 192.168.1.100 that is shared among the nodes, and one node always has it. This lets you use this IP in your kubeconfig file, too, to make the kube API server always reachable without any external dependencies. Or if you expose services to node ports, or just the 443 ingress port, you can always refer to 192.168.1.100:443 or an imaginary 192.168.1.100:45678 with any TCP traffic to reach a cluster node that is up. I hope it helps! :)

**Comment by fabiorauber** [comment]
Thanks for the tip @immanuelfodor.  I've been using a Keepalived container (https://github.com/interlegis/docker-keepalived) for this job for quite a while, but I believe a solution with MetalLB would be more nicely integrated and would provide a better experience. Nonetheless, it's nice to know we have alterenatives! :)

**Comment by asemen2608** [comment]
Please provide eta or Rancher version where RFE could be added

**Comment by Aaron-ML** [comment]
This would be a great add. I personally think rancher is incredibly well suited in on-premise environments, Anything that makes it easier to adopt is a win in my eyes.
